Add new module for Redfish APIs

jose-delarosa opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Module to automate Out-Of-Band (OOB) management using Redfish APIs available in OOB controllers such as Dell's iDRAC and HP's iLO.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
redfish

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.0a2 (detached HEAD 707f069fe8) last updated 2018/06/07 22:10:55 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /share1/git/ansible/lib/ansible
  executable location = /share1/git/ansible/bin/ansible
  python version = 2.7.5 (default, Feb 20 2018, 09:19:12) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]

```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->
The DMTF will extend functionality and maintain this community module.

<!--- Paste verbatim command output below, e.g. before and after your change -->

Commit:
Add new module for Redfish APIs

Commit:
PEP8 and other sanity check fixes

Commit:
Python 3.x sanity check fixes

jose-delarosa commented:
ready_for_review

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+       user={{user}} password={{password}} userid={{userid}}
+
+  - name: Get BIOS attributes
+    local_action: >
```
Use YAML form:

```
  redfish:
     category: Accounts
     command: GetBiosAttributes
     ...
```

jose-delarosa replied:
Fixed playbooks a while ago but forgot the examples. Done.

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+    rf_utils = RedfishUtils(creds, root_uri)
+
+    # Organize by Categories / Commands
+    if category == "Inventory":
```
recommend having a [category][command] invocation dict

```
invocation = dict(
    Inventory=dict(
        GetSystemInventory=rf_utils.get_system_inventory,
        ...
    )
    ...
)

result = find_resource(category, rf_uri)
if not result['ret']:
    module.fail_json(msg=result['msg'])
invocation[category][command]()
```

jose-delarosa replied:
I like this approach, much cleaner than all those if/else statements. However, not sure how straight-forward it can be applied here, as some conditionals execute additional checks before calling a function(). Let me take a look, but if not, can we label this as a nice-to-have or future improvement?

s-hertel commented on lib/ansible/module_utils/redfish_utils.py:
```diff
@@ -0,0 +1,940 @@
+        except URLError as e:
+            return {'ret': False, 'msg': "URL Error: %s" % e.reason}
+        except Exception as e:
+            return {'ret': False, 'msg': "Other error: %s" % e.message}
```
The e.message exception handling will cause an AttributeError on Python 3.

jose-delarosa replied:
Thank you. What do you suggest using instead? (I suppose a quick glance at the docs on my part wouldn't hurt).

juliakreger replied:
This might not work, but the __str__ return result of error might have what your looking for.

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+            user=dict(required=True, type='str'),
+            password=dict(required=True, type='str', no_log=True),
+            userid=dict(required=False, type='str'),
+            username=dict(required=False, type='str'),
```
`required=False` and `type='str'` are default, so no need to explicitly state them in argument_spec

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+'''
+
+RETURN = '''
+changed:
```
We don't usually add `changed` or `invocation` in `RETURN` as all modules return them

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+result:
+    description: different results depending on task
+    returned: always
+    type: string
```
Not a dict?

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+  category:
+    required: true
+    description:
+      - Action category to execute on server
```
Suggest adding `choices` to `category`, even if not to `command`

jose-delarosa replied:
Done in https://github.com/ansible/ansible/pull/41297/commits/5edc431e08450555660efff5b57a3978b380801c

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+version_added: "2.7"
+short_description: Out-Of-Band management using Redfish APIs
+description:
+  - Builds Redfish URIs locally and sends them to remote OOB controllers to
```
I would describe *what* this module does, rather than *how* it does it first.

> Manages out of band controllers using the redfish API

juliakreger replied:
+1, Think of the user as not being aware of how redfish works, but as someone who needs to accomplish a task.

jose-delarosa replied:
Makes sense, thank you both. Added to fix queue.

willthames commented on lib/ansible/modules/remote_management/redfish/redfish.py:
```diff
@@ -0,0 +1,331 @@
+    result = {}
+    module = AnsibleModule(
+        argument_spec=dict(
+            category=dict(required=True, type='str'),
```
Suggest adding `choices` to `category`, even if not to `command`

maxamillion commented:
ping @juliakreger - it was suggested that you would be a good candidate to review this code by @robynbergeron :)

juliakreger reviewed (commented):
It feels like this really should be broken into three separate chunks:

Something like "redfish_actions" to set/assert new power, "redfish_settings" to get/set bios/bmc settings, and "redfish_status" to return information as a dictionary that a user can use in their playbooks to make decisions upon the status of the BMC or what the BMC is reporting. I'm sure it can be broken up further but it really seems like someone would need to ether understand the contents of the module to use the module, and splitting it into smaller self contained modules would improve user experience and allow for more verbose documentation, so a systems administrator who is only aware that this "redfish" thing exists, can turn it on and quickly gain traction.

The logical separation would be useful for other major functional areas such as raid, virtual media, etc.

jose-delarosa commented:
@juliakreger, I had thought about breaking up as well for better manageability, but was thinking:
- redfish_facts - information/status is extracted (i.e. get hardware inventory, BIOS level)
- redfish_config - configure OOB controller (i.e. NTP settings, boot mode)
- redfish_command - command/action is run (i.e. power up/down)

Most current functionality falls under facts and config, but definitely should be easier to understand.

juliakreger commented:
@jose-delarosa That sounds perfect, although I would kind of suggest that boot mode instead be an action, because in some cases it can be a one-time boot setting, or a persistent setting for the running host, not the oob controller.  Then again, it could also be static config. I guess there are a few different ways that could be looked at. :)

Commit:
Updating examples to YAML format

Commit:
Removing required=False and type='str' as they are the default values.

Commit:
Updated short_description and RETURN example

Commit:
Added choices to category

Commit:
Modify exception handling to avoid AttributeError in python3

jose-delarosa commented:
@willthames, are you OK with changes so far? Still need to split the module as suggested by @juliakreger but want to make sure you're OK so far. As mentioned, I am not sure yet on the [category][command] invocation dict, though certainly something that could be added later on.

Thank you all for your reviews.

willthames commented:
@jose-delarosa yep, no major issues. You may want multiple facts modules as well as multiple action modules.

The [category][command] invocation stuff is really of most benefit for maintenance - as that's likely to fall on you anyway, it's totally up to you whether to implement it or not :)

jose-delarosa commented:
@s-hertel, I removed the e.message exception handling and put in a generic except. I believe URLError and HTTPError should catch most exceptions anyway. Thanks for your review.

Commit:
Split up module into 3 parts to improve user experience:

redfish_facts - extract inventory or status information
redfish_config - configure OOB controller settings
redfish_command - execute command or perform action

jose-delarosa commented:
Broke up into 3 modules but now have to follow process when submitting more than one module. Will submit PR for redfish_facts first and then follow up with other two.

abadger commented:
* Categories should probably be implemented as taking a list of categories.  That way the user can run redfish_facts once, rather than for each category they want to gather.  Among other things, that will be more efficient as the ansible module won't have to be transferred to the remote host and executed multiple times.
